File: wai_world_tools/picovoice_ros/picovoice_driver/extern/pvspeaker/demo/python/pv_speaker_demo.py
#
# Copyright 2024 Picovoice Inc.
#
# You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
# file accompanying this source.
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#


import argparse
import wave
import array

from pvspeaker import PvSpeaker


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "--show_audio_devices",
        help="List of audio devices currently available for use.",
        action="store_true")

    parser.add_argument(
        "--audio_device_index",
        help="Index of input audio device.",
        type=int,
        default=-1)

    parser.add_argument(
        "--input_wav_path",
        help="Path to PCM WAV file to be played.",
        default=None)

    args = parser.parse_args()

    if args.show_audio_devices:
        devices = PvSpeaker.get_available_devices()
        for i in range(len(devices)):
            print("index: %d, device name: %s" % (i, devices[i]))
    else:
        device_index = args.audio_device_index
        input_path = args.input_wav_path

        wavfile = None
        speaker = None

        try:
            if input_path is not None:
                wavfile = wave.open(input_path, "rb")

                sample_rate = wavfile.getframerate()
                bits_per_sample = wavfile.getsampwidth() * 8
                num_channels = wavfile.getnchannels()
                num_samples = wavfile.getnframes()

                if bits_per_sample != 8 and bits_per_sample != 16 and bits_per_sample != 24 and bits_per_sample != 32:
                    print(f"Unsupported bits per sample: {bits_per_sample}")
                    wavfile.close()
                    exit()

                if num_channels != 1:
                    print("WAV file must have a single channel (MONO)")
                    wavfile.close()
                    exit()

                speaker = PvSpeaker(
                    sample_rate=sample_rate,
                    bits_per_sample=bits_per_sample,
                    device_index=device_index)
                print("pvspeaker version: %s" % speaker.version)
                print("Using device: %s" % speaker.selected_device)

                wav_bytes = wavfile.readframes(num_samples)

                pcm = None
                if bits_per_sample == 8:
                    pcm = list(array.array('B', wav_bytes))
                elif bits_per_sample == 16:
                    pcm = list(array.array('h', wav_bytes))
                elif bits_per_sample == 24:
                    pcm = []
                    for i in range(0, len(wav_bytes), 3):
                        sample = int.from_bytes(wav_bytes[i:i + 3], byteorder='little', signed=True)
                        pcm.append(sample)
                elif bits_per_sample == 32:
                    pcm = list(array.array('i', wav_bytes))

                speaker.start()

                print("Playing audio...")
                speaker.write(pcm)
                speaker.stop()

                print("Finished playing audio...")
                wavfile.close()

        except KeyboardInterrupt:
            print("Stopping...")
        finally:
            if speaker is not None:
                speaker.delete()
            if wavfile is not None:
                wavfile.close()


if __name__ == "__main__":
    main()

File: wai_world_tools/picovoice_ros/picovoice_driver/extern/pvspeaker/demo/c/pv_speaker_demo.c
/*
    Copyright 2024 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
    file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
    specific language governing permissions and limitations under the License.
*/

#include <getopt.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "pv_speaker.h"

static volatile bool is_interrupted = false;

void interrupt_handler(int _) {
    (void) _;
    is_interrupted = true;
}

static struct option long_options[] = {
        {"show_audio_devices", no_argument,       NULL, 's'},
        {"input_wav_path",     required_argument, NULL, 'i'},
        {"audio_device_index", required_argument, NULL, 'd'}
};

static void print_usage(const char *program_name) {
    fprintf(stderr,
            "Usage : %s -i INPUT_WAV_PATH [-d AUDIO_DEVICE_INDEX]\n"
            "        %s --show_audio_devices\n",
            program_name,
            program_name);
}

static void show_audio_devices(void) {
    char **device_list = NULL;
    int32_t device_list_length = 0;

    // List devices
    pv_speaker_status_t status = pv_speaker_get_available_devices(
            &device_list_length,
            &device_list);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to get audio devices with: %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    fprintf(stdout, "Printing devices...\n");
    for (int32_t i = 0; i < device_list_length; i++) {
        fprintf(stdout, "index: %d, name: %s\n", i, device_list[i]);
    }

    pv_speaker_free_available_devices(device_list_length, device_list);
}

typedef struct {
    uint8_t chunk_id[4];
    uint32_t chunk_size;
    uint8_t format[4];
    uint8_t subchunk1_id[4];
    uint32_t subchunk1_size;
    uint16_t audio_format;
    uint16_t num_channels;
    uint32_t sample_rate;
    uint32_t byte_rate;
    uint16_t block_align;
    uint16_t bits_per_sample;
    uint8_t subchunk2_id[4];
    uint32_t subchunk2_size;
} wav_header;

void *read_wav_file(const char *filename, uint32_t *num_samples, uint32_t *sample_rate, uint16_t *bits_per_sample) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Unable to open file");
        return NULL;
    }

    wav_header header;

    fread(&header, sizeof(header), 1, file);

    if (header.chunk_id[0] != 'R' || header.chunk_id[1] != 'I' || header.chunk_id[2] != 'F' || header.chunk_id[3] != 'F' ||
        header.format[0] != 'W' || header.format[1] != 'A' || header.format[2] != 'V' || header.format[3] != 'E') {
        fclose(file);
        fprintf(stderr, "Invalid WAV file\n");
        return NULL;
    }

    if (header.audio_format != 1) {
        fclose(file);
        fprintf(stderr, "WAV file format must be PCM type\n");
        return NULL;
    }

    if (header.num_channels != 1) {
        fclose(file);
        fprintf(stderr, "WAV file must have a single channel (MONO)\n");
        return NULL;
    }

    *sample_rate = header.sample_rate;

    *bits_per_sample = header.bits_per_sample;

    uint32_t bytes_per_sample = header.bits_per_sample / 8;
    *num_samples = header.subchunk2_size / bytes_per_sample;

    void *pcm_data = malloc(header.subchunk2_size);
    if (!pcm_data) {
        perror("Memory allocation failed");
        fclose(file);
        return NULL;
    }

    fread(pcm_data, header.subchunk2_size, 1, file);

    fclose(file);

    return pcm_data;
}

int main(int argc, char *argv[]) {
    const char *input_wav_path = NULL;
    int32_t device_index = -1;

    int c;
    while ((c = getopt_long(argc, argv, "si:d:", long_options, NULL)) != -1) {
        switch (c) {
            case 's':
                show_audio_devices();
                return 0;
            case 'i':
                input_wav_path = optarg;
                break;
            case 'd':
                device_index = (int32_t) strtol(optarg, NULL, 10);
                break;
            default:
                exit(1);
        }
    }

    if (!input_wav_path) {
        print_usage(argv[0]);
        exit(1);
    }

    signal(SIGINT, interrupt_handler);
    fprintf(stdout, "pv_speaker version: %s\n", pv_speaker_version());

    uint32_t num_samples, sample_rate;
    uint16_t bits_per_sample;
    void *pcm = read_wav_file(input_wav_path, &num_samples, &sample_rate, &bits_per_sample);

    fprintf(stdout, "Initializing pv_speaker...\n");
    const int32_t frame_length = 512;
    pv_speaker_t *speaker = NULL;
    pv_speaker_status_t status = pv_speaker_init(
            sample_rate,
            frame_length,
            bits_per_sample,
            device_index,
            10,
            &speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to initialize device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    pv_speaker_set_debug_logging(speaker, true);

    const char *selected_device = pv_speaker_get_selected_device(speaker);
    fprintf(stdout, "Selected device: %s.\n", selected_device);

    status = pv_speaker_start(speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to start device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    fprintf(stdout, "Playing audio...\n");
    if (pcm) {
        char *pcmData = (char *) pcm;
        for (int i = 0; i < num_samples; i += frame_length) {
            bool is_last_frame = i + frame_length >= num_samples;

            status = pv_speaker_write(
                    speaker,
                    is_last_frame ? num_samples - i : frame_length,
                    &pcmData[i * bits_per_sample / 8]);
            if (status != PV_SPEAKER_STATUS_SUCCESS) {
                fprintf(stderr, "Failed to write with %s.\n", pv_speaker_status_to_string(status));
                exit(1);
            }
            if (is_interrupted) {
                fprintf(stdout, "\nStopped audio...\n");
                break;
            }
        }

        free(pcm);
    }

    status = pv_speaker_stop(speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to stop device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    if (!is_interrupted) {
        fprintf(stdout, "Finished playing audio...\n");
    }

    fprintf(stdout, "Deleting pv_speaker...\n");
    pv_speaker_delete(speaker);

    return 0;
}
